### ISSUE TITLE
--exclude-resources is not respected

### ISSUE THREAD

**Issue by suleymanakbas91** [vmware-tanzu/velero#2374]
I'm trying to upgrade our Velero chart on Kyma from 1.2.0 to 1.3.1 [(#7699)](https://github.com/kyma-project/kyma/pull/7699). I realized that exclude resources logic no longer works.

To restore from a backup, we run:

```
velero restore create \
    --from-backup "${BACKUP_NAME}" \
    --exclude-resources clusterservicebrokers.servicecatalog.k8s.io,clusterserviceclasses.servicecatalog.k8s.io,clusterserviceplans.servicecatalog.k8s.io,servicebindings.servicecatalog.k8s.io,servicebrokers.servicecatalog.k8s.io,serviceclasses.servicecatalog.k8s.io,serviceinstances.servicecatalog.k8s.io,serviceplans.servicecatalog.k8s.io \
    --restore-volumes \
    --wait
```

Although we exclude all those resources, I see that they are still being restored:

[restore.log](https://github.com/vmware-tanzu/velero/files/4396754/restore.log)

We also have a restore plugin that was working well with Velero 1.2.0, but it no longer works with 1.3.1. When I checked the logs, I realized that exclude resources was not respected there, either. 

[Here](https://github.com/kyma-project/kyma/blob/master/components/backup-plugins/internal/plugins/ignore_by_label.go#L35) we exclude `trigger.eventing.knative.dev` from the resources this plugin applies to, but it still applies the action to this resource.

Logs from Velero still has: 

`Ignoring restore of Trigger.eventing.knative.dev eventmeshtest-backup-test/event-mesh-trigger as it had label eventing.knative.dev/broker:default`

[velero-pod.log](https://github.com/vmware-tanzu/velero/files/4396753/velero-pod.log)

**Environment:**

- Velero version (use `velero version`): 1.3.1
- Kubernetes version (use `kubectl version`): 1.15.9-gke.26
- Cloud provider or hardware configuration: GKE

**Comment by skriss** [comment]
@suleymanakbas91 are all of these resources Custom Resources, for which you're also restoring the Custom Resource Definition as part of the same restore?

I'm wondering if the resources that you're trying to exclude are not available via the discovery API at the start of the restore, so can't be resolved properly at that point.

If that's the case, there are some more recent code changes in master that may help out, although the plugin issue may still remain.

**Comment by suleymanakbas91** [comment]
@skriss yes, that's the case. Are those changes included in 1.3.2?

**Comment by skriss** [comment]
@suleymanakbas91 unfortunately they are not. The relevant PR is https://github.com/vmware-tanzu/velero/pull/2248 which is in `master` but hasn't been cherrypicked into a patch release.

**Comment by skriss** [comment]
> We also have a restore plugin that was working well with Velero 1.2.0, but it no longer works with 1.3.1. When I checked the logs, I realized that exclude resources was not respected there, either.
>
> Here we exclude trigger.eventing.knative.dev from the resources this plugin applies to, but it still applies the action to this resource.

I'd be surprised if this behavior changed from v1.2 to v1.3. As far as I know, the issue here is that we try to resolve the plugin's includes/excludes via the discovery API before starting the restore. If those includes/excludes refer to a CRD that does not yet exist in the cluster, but rather will be created as part of the restore, then the "resolution" step fails and basically ignores that exclude. This behavior has always existed.

One workaround option would be to do two consecutive restores: in the first restore, restore *only* CRDs; in the second restore, restore everything else. That way, during the second restore, the CRDs will already exist so the discovery API resolution should succeed.

We can also look at delaying the resolution of a RestoreItemAction's include/exclude resources to later, so that it's possible to reference any newly-created CRDs in those.

**Comment by ashish-amarnath** [comment]
@suleymanakbas91 The change Steve was referring to should be in our 1.4 release. Please give that try and let us know if you continue to see this problem.

**Comment by suleymanakbas91** [comment]
Thanks for the heads up @ashish-amarnath, but I don't have the environment to test it anymore. Feel free to close the issue, if you believe the problem has been solved.

**Comment by skriss** [comment]
I believe that the issue with `--exclude-resources` on `velero restore create` has been resolved.

However, the issue with restore item action excluded resources likely still exists. I'm going to file a separate issue specifically for that, and will close out this parent issue.

**Comment by skriss** [comment]
Opened #2545, closing this one out.
